idb/migration/migration_test.go

```diff
@@ -329,3 +330,105 @@ func TestSuccessfulMigration(t *testing.T) {
 		})
 	}
 }
+
+// TestAvailabilityChannelCloses tests that the migration object closes the availability
+// channel when blocking migrations finish. If it is not the case, this test will deadlock
+// (and timeout).
+func TestAvailabilityChannelCloses(t *testing.T) {
+	// Migration 2 reads on this channel.
+	migrationTwoChannel := make(chan struct{})
+
+	tasks := []Task{
+		{
+			MigrationID: 1,
+			Handler: func() error {
+				return nil
+			},
+			DBUnavailable: true,
+		},
+		{
+			MigrationID: 2,
+			Handler: func() error {
+				<-migrationTwoChannel
+				return nil
+			},
+		},
+	}
+
+	m, err := MakeMigration(tasks, nil)
+	require.NoError(t, err)
+
+	availableCh := m.RunMigrations()
+	_, ok := <-availableCh
+	assert.False(t, ok)
+
+	// This will block until migration 1 reads from the channel.
+	migrationTwoChannel <- struct{}{}
+}
+
+// TestAvailabilityChannelClosesNoMigrations tests that the migration object closes
+// the availability channel when last migration, which is blocking, finishes.
+// If it is not the case, this test will deadlock (and timeout).
+func TestAvailabilityChannelClosesNoMigrations(t *testing.T) {
+	tasks := []Task{
+		{
+			MigrationID: 1,
+			Handler: func() error {
+				return nil
+			},
+			DBUnavailable: true,
+		},
+	}
+
+	m, err := MakeMigration(tasks, nil)
+	require.NoError(t, err)
+
+	availableCh := m.RunMigrations()
+	_, ok := <-availableCh
+	assert.False(t, ok)
+}
+
+// TestAvailabilityChannelClosesBlockingMigrationLast tests that the migration object
+// closes when there are no migrations. If it is not the case, this test will
+// deadlock (and timeout).
+func TestAvailabilityChannelClosesBlockingMigrationLast(t *testing.T) {
+	m, err := MakeMigration([]Task{}, nil)
+	require.NoError(t, err)
+
+	availableCh := m.RunMigrations()
+	_, ok := <-availableCh
+	assert.False(t, ok)
+}
+
+// TestAvailabilityChannelDoesNotCloseEarly tests that the migration object closes the availability
+// channel only after all blocking migrations are run.
+func TestAvailabilityChannelDoesNotCloseEarly(t *testing.T) {
+	tasks := []Task{
+		{
+			MigrationID: 1,
+			Handler: func() error {
+				return nil
+			},
+		},
+		{
+			MigrationID: 2,
+			Handler: func() error {
+				time.Sleep(10 * time.Millisecond)
+				return nil
+			},
+			DBUnavailable: true,
+		},
+	}
+
+	m, err := MakeMigration(tasks, nil)
+	require.NoError(t, err)
+
+	availableCh := m.RunMigrations()
+
+	time.Sleep(5 * time.Millisecond)
```

Comment:
Leave out `default` so that select will do the blocking, you can also include a timeout so that failures are detected faster.

Reply:
done, this is a bit nicer